packages/react-data-grid/src/HeaderRow.tsx

```diff
@@ -0,0 +1,157 @@
+import React from 'react';
+import shallowEqual from 'shallowequal';
+
+import HeaderCell from './HeaderCell';
+import getScrollbarSize from './getScrollbarSize';
+import { getColumn, getSize, isFrozen } from './ColumnUtils';
+import SortableHeaderCell from './common/cells/headerCells/SortableHeaderCell';
+import FilterableHeaderCell from './common/cells/headerCells/FilterableHeaderCell';
+import { HeaderRowType, HeaderCellType, DEFINE_SORT } from './common/enums';
+import { Column, ColumnList } from './common/types';
+
+export interface Props {
+  width?: number;
+  height: number;
+  columns: ColumnList;
+  onSort(): void;
+  onColumnResize(column: Column, width: number): void;
+  onColumnResizeEnd(column: Column, width: number): void;
+  style?: React.CSSProperties;
+  sortColumn?: string;
+  sortDirection?: DEFINE_SORT;
+  filterable?: boolean;
+  onFilterChange(): void;
+  onScroll?(): void;
+  rowType: HeaderRowType;
+  draggableHeaderCell: React.ComponentType<{ column: Column; onHeaderDrop(): void }>;
+  onHeaderDrop(): void;
+}
+
+export default class HeaderRow extends React.Component<Props> {
+  static displayName = 'HeaderRow';
+
+  private readonly cells: Array<HeaderCell | null> = [];
+
+  shouldComponentUpdate(nextProps: Props) {
+    return (
+      nextProps.width !== this.props.width
+      || nextProps.height !== this.props.height
+      || nextProps.columns !== this.props.columns
+      || !shallowEqual(nextProps.style, this.props.style)
+      || this.props.sortColumn !== nextProps.sortColumn
+      || this.props.sortDirection !== nextProps.sortDirection
+    );
+  }
+
+  getHeaderCellType(column: Column): HeaderCellType {
+    if (column.filterable && this.props.filterable) {
+      return HeaderCellType.FILTERABLE;
+    }
+
+    if (column.sortable && this.props.rowType !== HeaderRowType.FILTER) {
+      return HeaderCellType.SORTABLE;
+    }
+
+    return HeaderCellType.NONE;
+  }
+
+  getFilterableHeaderCell(column: Column) {
+    const FilterRenderer = column.filterRenderer || FilterableHeaderCell;
+    return <FilterRenderer column={column} onChange={this.props.onFilterChange} />;
```

Comment:
Filter renderer needs `getValidFilterValues` prop otherwise it throws an error 

![image](https://user-images.githubusercontent.com/1452717/56245310-bcad9780-6064-11e9-8f12-b24180e7db7e.png)

Eventually, I think we should remove `getValidFilterValues` prop as the component can fetch valid values on mount. Wdyt?

Reply:
I've added the prop, but I don't think this is something that should be passed as a prop to ReactDataGrid, as it's only used for custom filter renderers anyway.